tests/end_to_end_test/snapshot_create_test.go

```diff
@@ -113,3 +118,292 @@ func TestSnapshottingCacheDirectory(t *testing.T) {
+				{Name: "foo/bar/file2.txt"},
+				{Name: "foo/file1.txt"},
+				{Name: "foo/file2.txt"},
+				{Name: "file1.txt"},
+				{Name: "file2.txt"},
+			},
+			expected: []string{},
+		},
+		// {
+		// 	desc: "ignore_all_but_text_files",
+		// 	files: []testFileEntry{
+		// 		{
+		// 			Name: "/.kopiaignore",
+		// 			Content: []string{
+		// 				"*",
+		// 				"!*.txt",
+		// 				"!*/",
+		// 			},
+		// 		},
+		// 		{Name: "foo/bar/file.txt"},
+		// 		{Name: "foo/bar/file.png"},
+		// 		{Name: "foo/file.txt"},
+		// 		{Name: "foo/file.jpg"},
+		// 		{Name: "car/car.jpg"},
+		// 		{Name: "file.txt"},
+		// 		{Name: "file.bmp"},
+		// 	},
+		// 	expected: []string{
+		// 		"foo/",
+		// 		"foo/bar/",
+		// 		"car/", // all files in this are ignored, but the directory is still included.
+		// 		"foo/bar/file.txt",
+		// 		"foo/file.txt",
+		// 		"file.txt",
+		// 	},
+		// },
+		{
+			desc: "ignore_rooted_vs_unrooted_1",
+			files: []testFileEntry{
+				{
+					Name: "/.kopiaignore",
+					Content: []string{
+						"/A/",
+					},
+				},
+				{Name: "file.txt"},
+				{Name: "/A/file.txt"},
+				{Name: "/A/AA/file.txt"},
+				{Name: "/A/B/A/file.txt"},
+				{Name: "/B/A/file.txt"},
+			},
+			expected: []string{
+				".kopiaignore",
+				"file.txt",
+				"B/A/file.txt",
+			},
+		},
+		{
+			desc: "ignore_rooted_vs_unrooted_2",
+			files: []testFileEntry{
+				{
+					Name: "/.kopiaignore",
+					Content: []string{
+						"A/",
+					},
+				},
+				{Name: "file.txt"},
+				{Name: "/A/file.txt"},
+				{Name: "/A/AA/file.txt"},
+				{Name: "/A/B/A/file.txt"},
+				{Name: "/B/A/file.txt"},
+			},
+			expected: []string{
+				".kopiaignore",
+				"file.txt",
+				"B/", // directory is empty because all contents are ignored, but the empty directory is still included.
+			},
+		},
+		{
+			desc: "ignore_rooted_vs_unrooted_3",
+			files: []testFileEntry{
+				{
+					Name: "/.kopiaignore",
+					Content: []string{
+						"B/A/",
+					},
+				},
+				{Name: "file.txt"},
+				{Name: "/A/file.txt"},
+				{Name: "/A/AA/file.txt"},
+				{Name: "/A/B/A/file.txt"},
+				{Name: "/B/A/file.txt"},
+			},
+			expected: []string{
+				".kopiaignore",
+				"file.txt",
+				"A/file.txt",
+				"A/AA/file.txt",
+				"A/B/A/file.txt",
+				"B/", // directory is empty because all contents are ignored, but the empty directory is still included.
+			},
+		},
+		{
+			desc: "ignore_rooted_vs_unrooted_4",
+			files: []testFileEntry{
+				{
+					Name: "/.kopiaignore",
+					Content: []string{
+						"/B/A/",
+					},
+				},
+				{Name: "file.txt"},
+				{Name: "/A/file.txt"},
+				{Name: "/A/AA/file.txt"},
+				{Name: "/A/B/A/file.txt"},
+				{Name: "/B/A/file.txt"},
+			},
+			expected: []string{
+				".kopiaignore",
+				"file.txt",
+				"A/file.txt",
+				"A/AA/file.txt",
+				"A/B/A/file.txt",
+				"B/", // directory is empty because all contents are ignored, but the empty directory is still included.
+			},
+		},
+		{
+			desc: "ignore_rooted_vs_unrooted_5",
```

Comment:
This looks great! Can you also add couple more tests where there are multiple `.kopiaignore` in various locations with overlapping rules:

* `./.kopiaignore` and `./A/.kopiaignore`
* `./.kopiaignore`, `./A/.kopiaignore` and `./A/B/.kopiaignore`
* `./A/.kopiaignore` and `./A/B/.kopiaignore`
* `./A/.kopiaignore` and `./B/.kopiaignore`

to make sure that those get applied correctly.

Reply:
Sure, we can add some of these. But I think most of these could be handled by the tests of ignorefs rather than end-to-end tests?  They are much faster to run anyway.

Reply:
Good idea by the way! 👍   I added some tests (commented in this PR) which led me to discovered a bug in the implementation in #773 when an ignore file in a subdirectory negated a pattern in the parent directory.